velox/functions/prestosql/aggregates/ChecksumAggregate.cpp

```diff
@@ -0,0 +1,225 @@
+      VELOX_NYI("Complex Types not yet supported!");
+    }
+
+    if (decoded.isConstantMapping()) {
+      auto isNull = decoded.isNullAt(0);
+
+      if (!isNull) {
+        rows.template applyToSelected(
+            [&](vector_size_t row) { computeHash(groups[row], decoded, 0); });
+      } else {
+        rows.template applyToSelected(
+            [&](vector_size_t row) { computeHashForNull(groups[row]); });
+      }
+    } else {
+      rows.template applyToSelected([&](vector_size_t row) {
+        if (decoded.isNullAt(row)) {
+          computeHashForNull(groups[row]);
+        } else {
+          computeHash(groups[row], decoded, row);
+        }
+      });
+    }
+  }
+
+  void addIntermediateResults(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /**/) override {
+    VELOX_CHECK_EQ(args[0]->encoding(), VectorEncoding::Simple::FLAT);
+
+    auto vector = args[0]->asUnchecked<FlatVector<int64_t>>();
+    auto rawValues = vector->rawValues();
+
+    rows.applyToSelected(
+        [&](vector_size_t i) { *value<int64_t>(groups[i]) = rawValues[i]; });
+  }
+
+  void addSingleGroupRawInput(
+      char* group,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /**/) override {
+    DecodedVector decoded(*args[0], rows);
+
+    if (decoded.isConstantMapping()) {
+      auto selected = rows.countSelected();
+      if (!decoded.isNullAt(0)) {
+        computeHash(group, decoded, 0, selected);
+      } else {
+        computeHashForNull(group, selected);
+      }
+    } else {
+      rows.template applyToSelected([&](vector_size_t row) {
+        if (decoded.isNullAt(row)) {
+          computeHashForNull(group);
+        } else {
+          computeHash(group, decoded, row);
+        }
+      });
+    }
+  }
+
+  void addSingleGroupIntermediateResults(
+      char* group,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /**/) override {
+    VELOX_CHECK_EQ(args[0]->encoding(), VectorEncoding::Simple::FLAT);
+
+    auto vector = args[0]->asUnchecked<FlatVector<int64_t>>();
+    auto rawValues = vector->rawValues();
+    int64_t result = 0;
+    if (vector->mayHaveNulls()) {
+      rows.applyToSelected([&](vector_size_t i) {
+        if (!vector->isNullAt(i)) {
+          result += rawValues[i];
+        }
+      });
+    } else {
+      rows.applyToSelected([&](vector_size_t i) { result += rawValues[i]; });
+    }
+
+    *value<int64_t>(group) = result;
+  }
+
+ private:
+  inline void computeHash(
+      char* group,
+      const DecodedVector& vector,
+      const vector_size_t& row,
+      const uint32_t times = 1) {
+    exec::Aggregate::clearNull(group);
+
+    auto hashVal = VELOX_DYNAMIC_TYPE_DISPATCH(
+        getHash, vector.base()->typeKind(), vector, row);
+    auto result = *exec::Aggregate::value<int64_t>(group);
+
+    result += times * hashVal * kPrime64;
```

Comment:
Might it overflow during the calculation? Or do we know it won't happen?

Reply:
Shouldnt matter, it acts as a ring.